### PR TITLE
preconditions for RuleBasedStateMachine

### PR DESCRIPTION
**Pull request by radix**
This adds a `@precondition` decorator that you can use on your `@rule`-decorated functions to add preconditions to them. Rules that don't match preconditions will not be considered as a "step" in the state machine.

TODO:

- [x] docstring updates
- [x] prose documentation
- [x] tests (I was pretty much just using my own code as a test case; need to actually implement in-hypothesis unit tests for this feature).

### CODE REVIEW

**Comment by radix** [comment]
@DRMacIver Can I get some feedback on this before I finish it up? I managed to maintain backwards compatibility with the multi-`@rule` usage. The API this provides is that you can put a `@precondition` decorator *after* (i.e. lexically before) any `@rule` invocation and it will add a precondition to that rule.

I still need to work on docstrings, unit tests, and prose docs. I also need to implement a more robust metadata-copying for function objects when decorating them. There's a `functools.wraps` thing in the stdlib, but it's garbage IMO; it doesn't handle the case of decorating non-functions, or functions defined as lambdas. I can implement something a little bit better than `functools.wraps` if you're okay with including it in hypothesis (I already know exactly what it would look like). Or maybe you already have something that I haven't found?

**Comment by DRMacIver** [comment]
So this might be syntactically weird, but how would you feel about:

```python
@rule(...).precondition(

)
def blah(self, ...):
    pass
```

?

I'm not totally wild about the order of decorators mattering, though not insufficiently wild about it to accept it.

RE functools.wraps: I think you want hypothesis.internal.reflection.proxies, however I don't think you should need it. It's OK to just return the original function object isn't it?

**Comment by radix** [comment]
> `@rule(...).precondition`

I'll try out implementing this.

> RE functools.wraps: I think you want hypothesis.internal.reflection.proxies, however I don't think you should need it. It's OK to just return the original function object isn't it?

I changed the way that rules were stored to essentially be a linked list of decorator function wrappers in order to maintain support for the multi-`@rule` things while also allowing `@precondition` decorators. Your suggestion for the different syntax may make this unnecessary; I'll see how it goes. Thanks for the response! I'll try to get this stuff done today.

**Comment by radix** [comment]
UGH. So, Python doesn't allow @foo().bar(). I forgot about this. It artificially limits the complexity of expressions you can use as decorators.

This was going to be so much nicer :(

**Comment by radix** [comment]
I think this is getting pretty close. I've added a docstring for `precondition`, added a section in the prose docs about using preconditions, and added some unit tests.

There are probably some little nitpicks remaining, but I think this is ready for an official review now.

File: tests/cover/test_stateful.py [hunk 1]
```diff
@@ -173,15 +173,45 @@ def __init__(self, value):
 class DepthMachine(RuleBasedStateMachine):
     charges = Bundle(u'charges')
 
-    @rule(targets=(charges,), child=charges)
-    @rule(targets=(charges,), child=none())
-    def charge(self, child):
-        return DepthCharge(child)
+    # double-rule is deprecated
+    with Settings(strict=False):
```
Comment (DRMacIver) [inline tests/cover/test_stateful.py]: I'd like to see a test that asserts this deprecation. There are a bunch of these in tests/cover/test_deprecation.py

File: tests/cover/test_stateful.py [hunk 2]
```diff
@@ -173,15 +173,45 @@ def __init__(self, value):
...
+        print(data)
+
+    rule1 = rule(data=just(u"rule1data"))(myfunc)
+    rule2 = rule(data=just(u"rule2data"))(myfunc)
+
+
+class PreconditionMachine(RuleBasedStateMachine):
+    num = 0
+
+    @rule()
```
Comment (DRMacIver) [inline tests/cover/test_stateful.py]: It's probably not really necessary, but I wonder if it would make sense to also have a rule in this test that resets num to zero. Right now the opportunities for the preconditions firing are a bit thin on the ground.

I don't actually expect this to catch anything but it would increase my confidence in the test as a reader.

**Comment by DRMacIver** [comment]
Other than slightly pedantic requests for more tests, LGTM.

**Comment by DRMacIver** [comment]
:+1:
